liquidity/liquidity.go

```diff
@@ -420,6 +510,71 @@ func (m *Manager) makeLoopOutRequest(suggestion *LoopOutRecommendation,
 	}
 }
 
+// calculateExpenditure calculates the total amount that has been spent by
+// automatically dispatched swaps. If there is an automatically dispatched swap
+// currently in progress, this function will fail with errAutoSwapInFlight
+// because we cannot accurately calculate off-chain fees for an in-flight
+// swap.
+func (m *Manager) calculateExpenditure(ctx context.Context,
+	loopOuts []*loopdb.LoopOut) (btcutil.Amount, error) {
+
+	var totalFees btcutil.Amount
+
+	for _, out := range loopOuts {
+		if out.Contract.Label != labels.AutoOutLabel() {
+			continue
+		}
+
+		// If the swap is pending, exit early with an error. We can't
+		// calculate total fees, and any effort to do so may race with
+		// the swap being updated after we have added this on-disk
+		// snapshot.
+		state := out.State().State.Type()
+		if state == loopdb.StateTypePending {
+			return 0, errAutoSwapInFlight
+		}
+
+		// We are not pending, so we know our last update is non-nil.
+		// Swaps that fall before our budget start date have no impact
+		// on our current budget calculation. We use update rather than
+		// creation time because fees are not immediately paid at our
+		// start time.
+		if out.LastUpdate().Time.Before(m.params.AutoFeeStartDate) {
+			continue
+		}
+
+		switch out.State().State.Type() {
+		// If we succeeded, we know that we have an non-nil last update
+		// so we can just add our total cost.
+		case loopdb.StateTypeSuccess:
+			totalFees += out.LastUpdate().Cost.Total()
+
+		// If our swap failed, we will have have paid fees. However, if
+		// an automated swap failed but the prepay was already pulled,
+		// we add that amount to our fee total, because it is an expense
+		// that the auto-looper incurred.
+		case loopdb.StateTypeFail:
+			if out.State().State == loopdb.StateFailSweepTimeout {
```

Comment:
Could make the `switch` statement on the `State` directly to avoid this nested `if`. Would almost look the same, except above it would be `loopdb.StateSuccess` instead of `loopdb.StateTypeSuccess`.